### ISSUE TITLE
False positive for SC1003 due to enclosing quotation

### ISSUE THREAD

**Issue by zackw** [koalaman/shellcheck#2872]
- Rule Id: SC1003
- My shellcheck version: 0.9.0
- [x] The rule's wiki page does not already cover this
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#! /bin/sh

trap '
  case $val in
    (*\'\''*) val=$(printf '\''%s\n'\'' "$val" | sed "s/'\''/'\''\\\\\\\\'\'''\''/g") ;;
  esac
  printf '\''%s\n'\'' "val='\''$val'\''"
' 0

val="$1"
```

This is cut down from logic actually used in Autoconf.

#### Here's what shellcheck currently says:

```
In test.sh line 5:
    (*\'\''*) val=$(printf '\''%s\n'\'' "$val" | sed "s/'\''/'\''\\\\\\\\'\'''\''/g") ;;
      ^-- SC1003 (info): Want to escape a single quote? echo 'This is how it'\''s done'.
                                                                        ^-- SC1003 (info): Want to escape a single quote? echo 'This is how it'\''s done'.
```

#### Here's what I wanted or expected to see:

No SC1003 complaints; this code is correct as-is.  It looks like shellcheck has gotten confused by the way an entire glob of shell code has been wrapped in single quotes so it can be used as an argument to 'trap'.  Yes, we probably ought to factor this out to a shell function, but it didn't used to be possible to use shell functions in Autoconf-generated configure scripts and I'm trying to put a release out so I don't want to make invasive changes right now.

**Comment by koalaman** [comment]
You're right, of course. ShellCheck does interpret it correctly, but generally warns whenever `\'` is found in a single quoted string because of the frequency of newbies thinking it'll behave like Python `'foo\'bar'`. Unfortunately I don't really see a reliable way to differentiate in such a way that they can get the support you need while it cleanly stays out of your way here.

Since you clearly know how to escape a complex string, please don't hesitate to globally ignore this error code via a directive. Obviously it's annoying to get linting warnings on perfectly correct code, but I hope you're willing to suffer it to help out less experienced users.
